File: src/rect.go
// rect.go
//
// Copyright (c) 2026 PDFjet Software
// Licensed under the MIT License. See LICENSE file in the project root.

package pdfjet

import (
	"github.com/edragoev1/pdfjet/v9/src/pathoperator"
	"github.com/edragoev1/pdfjet/v9/src/single"
)

// Rect is used to create rectangular shapes on a page.
type Rect struct {
	x              float32
	y              float32
	width          float32
	height         float32
	cornerRadius   float32
	fillColor      [3]float32
	hasFillColor   bool
	borderWidth    float32
	borderColor    [3]float32
	hasBorderColor bool
	borderPattern  string
	uri            string
	key            string
	language       string
	altDescription string
	actualText     string
}

// NewRect creates new Rect object.
// @param x the x coordinate of the top left corner of this rect when drawn on the page.
// @param y the y coordinate of the top left corner of this rect when drawn on the page.
// @param w the width of this rect.
// @param h the height of this rect.
func NewRect(x, y, w, h float32) *Rect {
	rect := new(Rect)
	rect.x = x
	rect.y = y
	rect.width = w
	rect.height = h

	rect.borderWidth = 0.0
	rect.borderPattern = "[] 0"

	rect.altDescription = single.Space
	rect.actualText = single.Space
	return rect
}

// SetLocation sets the location of this rect on the page.
// @param x the x coordinate of the top left corner of this rect when drawn on the page.
// @param y the y coordinate of the top left corner of this rect when drawn on the page.
func (rect *Rect) SetLocation(x, y float32) Drawable {
	rect.x = x
	rect.y = y
	return rect
}

// SetSize sets the size of this rect.
// @param w the width of this rect.
// @param h the height of this rect.
func (rect *Rect) SetSize(w, h float32) *Rect {
	rect.width = w
	rect.height = h
	return rect
}

// SetBorderColor sets the border color as a 0xRRGGBB value.
func (rect *Rect) SetBorderColor(color int32) *Rect {
	r := float32((color>>16)&0xff) / 255.0
	g := float32((color>>8)&0xff) / 255.0
	b := float32((color)&0xff) / 255.0
	rect.SetBorderColorRGB([3]float32{r, g, b})
	return rect
}

// SetBorderColorRGB sets the color for this rectangle.
// @param color the color specified as an integer.
func (rect *Rect) SetBorderColorRGB(borderColor [3]float32) *Rect {
	rect.borderColor = borderColor
	rect.hasBorderColor = true
	return rect
}

// SetFillColor sets the fill color as a 0xRRGGBB value.
func (rect *Rect) SetFillColor(color int32) *Rect {
	r := float32((color>>16)&0xff) / 255.0
	g := float32((color>>8)&0xff) / 255.0
	b := float32((color)&0xff) / 255.0
	rect.SetFillColorRGB([3]float32{r, g, b})
	return rect
}

// SetFillColorRGB sets the fill color from red, green and blue values.
func (rect *Rect) SetFillColorRGB(fillColor [3]float32) *Rect {
	rect.fillColor = fillColor
	rect.hasFillColor = true
	return rect
}

// SetBorderWidth sets the width of this line.
// @param width the width.
func (rect *Rect) SetBorderWidth(borderWidth float32) *Rect {
	rect.borderWidth = borderWidth
	return rect
}

// SetBorderPattern sets the line dash pattern of the border.
func (rect *Rect) SetBorderPattern(borderPattern string) *Rect {
	rect.borderPattern = borderPattern
	return rect
}

// SetCornerRadius sets the corner radius.
// @param width the width.
func (rect *Rect) SetCornerRadius(cornerRadius float32) *Rect {
	rect.cornerRadius = cornerRadius
	return rect
}

// SetURIAction sets the URI for the "click rect" action.
// @param uri the URI
func (rect *Rect) SetURIAction(uri string) *Rect {
	rect.uri = uri
	return rect
}

// SetGoToAction sets the destination key for the action.
// @param key the destination name.
func (rect *Rect) SetGoToAction(key string) *Rect {
	rect.key = key
	return rect
}

// SetLanguage sets the language of this rect, used for accessibility.
// @param language the language, for example "en-US".
func (rect *Rect) SetLanguage(language string) *Rect {
	rect.language = language
	return rect
}

// SetAltDescription sets the alternate description of this rect.
// @param altDescription the alternate description of the rect.
// @return this Rect.
func (rect *Rect) SetAltDescription(altDescription string) *Rect {
	rect.altDescription = altDescription
	return rect
}

// SetActualText sets the actual text for this rect.
// @param actualText the actual text for the rect.
// @return this Rect.
func (rect *Rect) SetActualText(actualText string) *Rect {
	rect.actualText = actualText
	return rect
}

// ScaleBy scales this rect by the specified factor.
// @param factor the factor used to scale the rect.
func (rect *Rect) ScaleBy(factor float32) {
	rect.x *= factor
	rect.y *= factor
}

// DrawOn draws this rect on the specified page.
// @param page the page to draw this rect on.
// @return x and y coordinates of the bottom right corner of this component.
func (rect *Rect) DrawOn(page *Page) [2]float32 {
	const k float32 = 0.55228

	// A rectangle carries no text, so it is decorative content.
	page.AddArtifactBMC()
	page.SaveGraphicsState()
	if rect.cornerRadius == 0.0 {
		if rect.hasFillColor {
			page.MoveTo(rect.x, rect.y)
			page.LineTo(rect.x+rect.width, rect.y)
			page.LineTo(rect.x+rect.width, rect.y+rect.height)
			page.LineTo(rect.x, rect.y+rect.height)
			page.LineTo(rect.x, rect.y)
			page.SetBrushColorRGB(rect.fillColor)
			page.FillPath()
		}
		if rect.hasBorderColor {
			page.MoveTo(rect.x, rect.y)
			page.LineTo(rect.x+rect.width, rect.y)
			page.LineTo(rect.x+rect.width, rect.y+rect.height)
			page.LineTo(rect.x, rect.y+rect.height)
			page.SetPenColorRGB(rect.borderColor)
			page.SetPenWidth(rect.borderWidth)
			page.SetStrokeDashPattern(rect.borderPattern)
			page.ClosePath()
		}
	} else {
		// The pen and brush must be set before the path is painted,
		// otherwise the rounded rectangle is drawn with whatever state
		// the page happened to be left in.
		if rect.hasBorderColor {
			page.SetStrokeDashPattern(rect.borderPattern)
		}
		if rect.hasFillColor {
			page.SetBrushColorRGB(rect.fillColor)
		}
		if rect.hasBorderColor {
			page.SetPenWidth(rect.borderWidth)
			page.SetPenColorRGB(rect.borderColor)
		}

		points := make([]*Point, 0)
		points = append(points, NewPoint(rect.x+rect.cornerRadius, rect.y))
		points = append(points, NewPoint((rect.x+rect.width)-rect.cornerRadius, rect.y))
		points = append(points, NewControlPointC((rect.x+rect.width-rect.cornerRadius)+rect.cornerRadius*k, rect.y))
		points = append(points, NewControlPointC(rect.x+rect.width, (rect.y+rect.cornerRadius)-rect.cornerRadius*k))
		points = append(points, NewPoint(rect.x+rect.width, rect.y+rect.cornerRadius))
		points = append(points, NewPoint(rect.x+rect.width, (rect.y+rect.height)-rect.cornerRadius))
		points = append(points, NewControlPointC(rect.x+rect.width, ((rect.y+rect.height)-rect.cornerRadius)+rect.cornerRadius*k))
		points = append(points, NewControlPointC(((rect.x+rect.width)-rect.cornerRadius)+rect.cornerRadius*k, rect.y+rect.height))
		points = append(points, NewPoint((rect.x+rect.width)-rect.cornerRadius, rect.y+rect.height))
		points = append(points, NewPoint(rect.x+rect.cornerRadius, rect.y+rect.height))
		points = append(points, NewControlPointC((rect.x+rect.cornerRadius)-rect.cornerRadius*k, rect.y+rect.height))
		points = append(points, NewControlPointC(rect.x, ((rect.y+rect.height)-rect.cornerRadius)+rect.cornerRadius*k))
		points = append(points, NewPoint(rect.x, (rect.y+rect.height)-rect.cornerRadius))
		points = append(points, NewPoint(rect.x, rect.y+rect.cornerRadius))
		points = append(points, NewControlPointC(rect.x, (rect.y+rect.cornerRadius)-rect.cornerRadius*k))
		points = append(points, NewControlPointC((rect.x+rect.cornerRadius)-rect.cornerRadius*k, rect.y))
		points = append(points, NewPoint(rect.x+rect.cornerRadius, rect.y))

		if rect.hasFillColor && !rect.hasBorderColor {
			page.DrawPath(points, pathoperator.Fill)
		} else if !rect.hasFillColor && rect.hasBorderColor {
			page.DrawPath(points, pathoperator.Stroke)
		} else if rect.hasFillColor && rect.hasBorderColor {
			page.DrawPath(points, pathoperator.FillAndStroke)
		}
	}
	page.RestoreGraphicsState()
	page.AddEMC()

	if rect.uri != "" || rect.key != "" {
		page.addAnnotation(&Annotation{
			annotationType: AnnotationLink,
			x1:             rect.x,
			y1:             rect.y,
			x2:             rect.x + rect.width,
			y2:             rect.y + rect.height,
			vertices:       nil,
			fillColor:      [3]float32{1.0, 1.0, 1.0}, // White color
			transparency:   0.0,
			title:          "",
			contents:       "",
			uri:            rect.uri,
			key:            rect.key, // The destination name
			language:       rect.language,
			actualText:     rect.actualText,
			altDescription: rect.altDescription,
		})
	}

	return [2]float32{rect.x + rect.width, rect.y + rect.height}
}

File: net/pdfjet/Rect.cs
/*
 * Rect.cs
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
using System;
using System.Collections.Generic;

namespace PDFjet.NET {
/// <summary>A rectangle that can be drawn on a page.</summary>
public class Rect  : IDrawable {
    internal float x;
    internal float y;
    private float w;
    private float h;
    private float r;

    private float[] fillColor;
    private float[] borderColor;
    private float borderWidth;
    private string borderPattern = "[] 0";

    private string uri;
    private string key;
    private string language = null;
    private string actualText = null;
    private string altDescription = null;

    /// <summary>
    /// The default constructor.
    /// </summary>
    public Rect() {
    }

    /// <summary>Creates a rectangle with its top left corner at x, y and the specified width and height.</summary>
    public Rect(float x, float y, float w, float h) {
        this.x = x;
        this.y = y;
        this.w = w;
        this.h = h;
    }

    /// <summary>Creates a rectangle with its top left corner at x, y and the specified width and height.</summary>
    public Rect(double x, double y, double w, double h) {
        this.x = (float) x;
        this.y = (float) y;
        this.w = (float) w;
        this.h = (float) h;
    }

    /// <summary>Sets the location of the top left corner of this rectangle.</summary>
    public Rect SetLocation(float x, float y) {
        this.x = x;
        this.y = y;
        return this;
    }

    /// <summary>Sets the location of the top left corner of this rectangle.</summary>
    public Rect SetLocation(double x, double y) {
        return SetLocation((float) x, (float) y);
    }

    IDrawable IDrawable.SetLocation(float x, float y) {
        return SetLocation(x, y);
    }

    /// <summary>Sets the size of this rectangle.</summary>
    public Rect SetSize(float w, float h) {
        this.w = w;
        this.h = h;
        return this;
    }

    /// <summary>Sets the fill color as a 0xRRGGBB value.</summary>
    public Rect SetFillColor(int color) {
        float r = ((color >> 16) & 0xff)/255f;
        float g = ((color >>  8) & 0xff)/255f;
        float b = ((color)       & 0xff)/255f;
        SetFillColor(r, g, b);
        return this;
    }

    /// <summary>Sets the fill color from red, green and blue values between 0.0 and 1.0.</summary>
    public Rect SetFillColor(float r, float g, float b) {
        this.fillColor = new float[] {r, g, b};
        return this;
    }

    /// <summary>Sets the fill color from an array of red, green and blue values.</summary>
    public Rect SetFillColor(float[] rgbColor) {
        this.fillColor = rgbColor;
        return this;
    }

    /// <summary>Sets the border width.</summary>
    public Rect SetBorderWidth(float width) {
        this.borderWidth = width;
        return this;
    }

    /// <summary>Sets the border color as a 0xRRGGBB value. Color.transparent removes the border.</summary>
    public Rect SetBorderColor(int color) {
        if (color == Color.transparent) {
            this.borderColor = null;
            return this;
        }
        float r = ((color >> 16) & 0xff)/255f;
        float g = ((color >>  8) & 0xff)/255f;
        float b = ((color)       & 0xff)/255f;
        SetBorderColor(r, g, b);
        return this;
    }

    /// <summary>Sets the border color from red, green and blue values between 0.0 and 1.0.</summary>
    public Rect SetBorderColor(float r, float g, float b) {
        this.borderColor = new float[] {r, g, b};
        return this;
    }

    /// <summary>Sets the border color from an array of red, green and blue values, or null for no border.</summary>
    public Rect SetBorderColor(float[] rgbColor) {
        this.borderColor = rgbColor;
        return this;
    }

    /// <summary>Sets the corner radius.</summary>
    public Rect SetCornerRadius(float r) {
        this.r = r;
        return this;
    }

    /// <summary>Sets the URI opened when this rectangle is clicked.</summary>
    public Rect SetURIAction(string uri) {
        this.uri = uri;
        return this;
    }

    /// <summary>Sets the destination key used when this rectangle is clicked.</summary>
    public Rect SetGoToAction(string key) {
        this.key = key;
        return this;
    }

    /// <summary>Sets the language of this rectangle, used for accessibility. When it is not set, the language of the document is used.</summary>
    public Rect SetLanguage(String language) {
        this.language = language;
        return this;
    }

    /// <summary>Sets the actual text of this rectangle, used for accessibility.</summary>
    public Rect SetActualText(String actualText) {
        this.actualText = actualText;
        return this;
    }

    /// <summary>Sets the alternate description of this rectangle, used for accessibility.</summary>
    public Rect SetAltDescription(string altDescription) {
        this.altDescription = altDescription;
        return this;
    }

    /// <summary>Sets the dash pattern of the border.</summary>
    public Rect SetBorderPattern(String borderPattern) {
        this.borderPattern = borderPattern;
        return this;
    }

    /// <summary>Multiplies the x and y coordinates of this rectangle by the specified factor.</summary>
    public void ScaleBy(float factor) {
        this.x *= factor;
        this.y *= factor;
    }

    /// <summary>Draws this rectangle on the specified page.</summary>
    public float[] DrawOn(Page page) {
        if (page == null) {
            return new float[] {x + w, y + h};
        }

        // A rectangle carries no text, so it is decorative content.
        page.AddArtifactBMC();
        page.SaveGraphicsState();

        const float k = 0.55228f;
        if (this.r == 0.0f) {
            if (fillColor != null) {
                page.MoveTo(this.x, this.y);
                page.LineTo(this.x + this.w, this.y);
                page.LineTo(this.x + this.w, this.y + this.h);
                page.LineTo(this.x, this.y + this.h);
                page.LineTo(this.x, this.y);
                page.SetBrushColor(this.fillColor);
                page.FillPath();
            }
            if (borderColor != null) {
                page.MoveTo(this.x, this.y);
                page.LineTo(this.x + this.w, this.y);
                page.LineTo(this.x + this.w, this.y + this.h);
                page.LineTo(this.x, this.y + this.h);
                page.SetPenColor(this.borderColor);
                page.SetPenWidth(this.borderWidth);
                page.SetStrokeDashPattern(this.borderPattern);
                page.ClosePath();
            }
        } else {
            // The pen and brush must be set before the path is painted,
            // otherwise the rounded rectangle is drawn with whatever state
            // the page happened to be left in.
            if (borderColor != null && borderPattern != null) {
                page.SetStrokeDashPattern(borderPattern);
            }
            if (fillColor != null) {
                page.SetBrushColor(fillColor);
            }
            if (borderColor != null) {
                page.SetPenWidth(borderWidth);
                page.SetPenColor(borderColor);
            }

            List<Point> points = new List<Point> {
                new Point((this.x + this.r), this.y),
                new Point((this.x + this.w) - this.r, this.y),
                new Point((this.x + this.w - this.r) + this.r * k, this.y, Point.ControlPointC),
                new Point((this.x + this.w), (this.y + this.r) - this.r * k, Point.ControlPointC),
                new Point((this.x + this.w), (this.y + this.r)),
                new Point((this.x + this.w), (this.y + this.h) - this.r),
                new Point((this.x + this.w), ((this.y + this.h) - this.r) + this.r * k, Point.ControlPointC),
                new Point(((this.x + this.w) - this.r) + this.r * k, (this.y + this.h), Point.ControlPointC),
                new Point(((this.x + this.w) - this.r), (this.y + this.h)),
                new Point((this.x + this.r), (this.y + this.h)),
                new Point(((this.x + this.r) - this.r * k), (this.y + this.h), Point.ControlPointC),
                new Point(this.x, ((this.y + this.h) - this.r) + this.r * k, Point.ControlPointC),
                new Point(this.x, (this.y + this.h) - this.r),
                new Point(this.x, (this.y + this.r)),
                new Point(this.x, (this.y + this.r) - this.r * k, Point.ControlPointC),
                new Point((this.x + this.r) - this.r * k, this.y, Point.ControlPointC),
                new Point((this.x + this.r), this.y)
            };
            if (fillColor != null && borderColor == null) {
                page.DrawPath(points, PathOperator.Fill);
            } else if (fillColor == null && borderColor != null) {
                page.DrawPath(points, PathOperator.Stroke);
            } else if (fillColor != null && borderColor != null) {
                page.DrawPath(points, PathOperator.FillAndStroke);
            }
        }

        page.RestoreGraphicsState();
        page.AddEMC();

        if (this.uri != null || this.key != null) {
            page.AddAnnotation(new Annotation(
                Annotation.Link,
                this.x,
                this.y,
                this.x + this.w,
                this.y + this.h,
                null,       // Vertices
                null,       // Fill Color
                0f,         // Transparency
                null,       // Title
                null,       // Contents
                this.uri,
                this.key,
                this.language,
                this.actualText,
                this.altDescription
            ));
        }

        return new float[] { this.x + this.w, this.y + this.h };
    }
}
}
